data-prepper-plugins/user-agent-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/useragent/CaffeineCachingParser.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.processor.useragent;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import ua_parser.Client;
+import ua_parser.Device;
+import ua_parser.OS;
+import ua_parser.Parser;
+import ua_parser.UserAgent;
+
+import java.util.function.Function;
+
+/**
+ * A superclass of {@link Parser} which uses Caffeine as a cache.
+ */
+class CaffeineCachingParser extends Parser {
+    private final Cache<String, Client> clientCache;
+    private final Cache<String, UserAgent> userAgentCache;
+    private final Cache<String, Device> deviceCache;
+    private final Cache<String, OS> osCache;
+
+    /**
+     * Constructs a new instance with a given cache size.
+     *
+     * @param cacheSize The size of the cache as a count of items.
+     */
+    CaffeineCachingParser(final long cacheSize) {
+        userAgentCache = createCache(cacheSize);
+        clientCache = createCache(cacheSize);
+        deviceCache = createCache(cacheSize);
+        osCache = createCache(cacheSize);
+    }
+
+    @Override
+    public Client parse(final String agentString) {
+        return parseCaching(agentString, clientCache, super::parse);
+    }
+
+    @Override
+    public UserAgent parseUserAgent(final String agentString) {
+        return parseCaching(agentString, userAgentCache, super::parseUserAgent);
+    }
+
+    @Override
+    public Device parseDevice(final String agentString) {
+        return parseCaching(agentString, deviceCache, super::parseDevice);
+    }
+
+    @Override
+    public OS parseOS(final String agentString) {
+        return parseCaching(agentString, osCache, super::parseOS);
+    }
+
+    private <T> T parseCaching(
+            final String agentString,
+            final Cache<String, T> cache,
+            final Function<String, T> parseFunction) {
+        if (agentString == null) {
+            return null;
+        }
```

Comment:
Is this different or faster from what Caffeine would return?

Reply:
This is what the `CachingParser` does already.

https://github.com/ua-parser/uap-java/blob/61ffd25de944bbe634ff5180104c9cdc07faad29/src/main/java/ua_parser/CachingParser.java#L62-L64

Also, we can avoid storing `null` in the cache.

Reply:
Caffeine does not store null values: https://github.com/ben-manes/caffeine/issues/118#issuecomment-243057017. This is also my experience with it. In one of my projects I had to work around this. But here we are checking the key not to be null, right? If we didn't, this would create an NPE. So I think, this is totally fine.